src/lib/swap-router/utils/trade.utils.ts

```diff
@@ -0,0 +1,100 @@
+import { BigNumber } from 'bignumber.js';
+
+import { RoutePairWithDirection } from '../interface/route-pair-with-direction.interface';
+import { Trade, TradeOperation } from '../interface/trade.interface';
+import { getPairFeeRatio } from './fee.utils';
+
+const findSwapOutput = (aTokenAmount: BigNumber, pair: RoutePairWithDirection) => {
+  const feeRatio = getPairFeeRatio(pair);
+
+  const aTokenAmountWithFee = aTokenAmount.times(feeRatio);
+
+  const numerator = aTokenAmountWithFee.times(pair.bTokenPool);
+  const denominator = pair.aTokenPool.plus(aTokenAmountWithFee);
+
+  return numerator.idiv(denominator);
+};
+
+const findSwapInput = (bTokenAmount: BigNumber, pair: RoutePairWithDirection) => {
+  const feeRatio = getPairFeeRatio(pair);
+
+  const numerator = pair.aTokenPool.times(bTokenAmount);
+  const denominator = pair.bTokenPool.minus(bTokenAmount).times(feeRatio);
+
+  const input = numerator.idiv(denominator).plus(1);
+
+  return input.isGreaterThan(0) ? input : new BigNumber(Infinity);
+};
+
+const getTradeOperationExactInput = (
+  aTokenAmount: BigNumber,
+  pair: RoutePairWithDirection,
+  slippageToleranceRatio: number
+): TradeOperation => ({
+  ...pair,
+  aTokenAmount,
+  bTokenAmount: findSwapOutput(aTokenAmount.multipliedBy(slippageToleranceRatio), pair)
+});
+
+const getTradeOperationExactOutput = (bTokenAmount: BigNumber, pair: RoutePairWithDirection): TradeOperation => ({
+  ...pair,
+  bTokenAmount,
+  aTokenAmount: findSwapInput(bTokenAmount, pair)
+});
+
+export const calculateTradeExactInput = (
+  inputAssetAmount: BigNumber,
+  routePairs: RoutePairWithDirection[],
+  slippageTolerancePercent = 0
+) => {
+  const trade: Trade = [];
+
+  if (routePairs.length > 0) {
+    const slippageToleranceRatio = (100 - slippageTolerancePercent) / 100;
```

Comment:
seems like its very common operation (third file), maybe consider to extract to helper/util function?

Reply:
whole `lib/swap-router` will be moved to the external library in the next PR (shared with QS), it could not depend on `temple` utils.  
There will be only 1 place where this util will be used, so I think there are no sense of doing that.